### ISSUE TITLE
General Fermi Hubbard model switches x and y dimensions

### ISSUE THREAD

**Issue by kevinsung** [quantumlib/OpenFermion#492]
I just noticed that in `hamiltonians/_general_hubbard_test.py` there is a test
```
def test_fermi_hubbard_square_special_general_equivalence(
        x_dimension, y_dimension, tunneling, coulomb,
        chemical_potential, spinless, periodic, magnetic_field):
    hubbard_model_special = fermi_hubbard(
            y_dimension, x_dimension, tunneling, coulomb,
            chemical_potential=chemical_potential, spinless=spinless,
            periodic=periodic, magnetic_field=magnetic_field)
    hubbard_model_general = fermi_hubbard_from_general(
            x_dimension, y_dimension, tunneling, coulomb,
            chemical_potential=chemical_potential, spinless=spinless,
            periodic=periodic, magnetic_field=magnetic_field)
    assert hubbard_model_special == hubbard_model_general
```
as you can see in the two calls, the order of `y_dimension` and `x_dimension` is swapped. If you make them match then the test fails. @bryano do you know what's up with that?

**Comment by bryano** [comment]
I used the opposite convention for linearizing the indices. The swapping in the test was intentional.

**Comment by kevinsung** [comment]
Oh ... I would have given you a hard time about that if I had noticed it before. Is there a particular reason you changed the convention? I think we should be consistent.

**Comment by bryano** [comment]
I thought I had a reason, but I don't what it is at this point. What do you think about having `_hubbard.py` just use `_general_hubbard.py`? The former is just a special case of the latter, but I imagine people have code depending on its interface, so we could just turn it into a wrapper around the general case.

In any case, changing the index linearization convention is just a matter of changing two lines in `lattice.py`.

**Comment by kevinsung** [comment]
I'd rather keep them separate for now. I've been brewing on some more general design changes that might affect all of this.

We should change the indexing convention to match with the original function.

**Comment by bryano** [comment]
I realize the reason now: my convention is consistent with `itertools.product`, in the sense that 
```
lattice.to_site_index((x, y) for x, y in itertools.product(range(x_dimension), range(y_dimension))
```
iterates over `range(x_dimension * y_dimension)`.

Are you okay with my changing `_hubbard.py` to my convention instead of vice versa?

**Comment by kevinsung** [comment]
No, please change the newer code to use the old convention. This isn't a good enough reason to break people's code over.

**Comment by bryano** [comment]
#493

**Comment by kevinsung** [comment]
Fixed in #493
